### DISCUSSION TITLE
아래 모듈 코드가 평가, 실행되는 과정을 설명해주세요.

### DISCUSSION THREAD

**Discussion by publdaze** [publdaze/js-study#89] [클로저]
- 아래 모듈 코드가 평가, 실행되는 과정을 설명해주세요.
  ```js
  // myModule.js
  let x = 5;
  export const getX = () => x;
  export const setX = (val) => {
    x = val;
  };
  
  // --- 파일 분리 ---
  
  import { getX, setX } from "./myModule.js";
  
  console.log(getX());
  setX(6);
  console.log(getX());
  ```

**Comment by publdaze** [comment]
> * 아래 모듈 코드가 평가, 실행되는 과정을 설명해주세요.
>   ```js
>   // myModule.js
>   let x = 5;
>   export const getX = () => x;
>   export const setX = (val) => {
>     x = val;
>   };
>   
>   // --- 파일 분리 ---
>   
>   import { getX, setX } from "./myModule.js";
>   
>   console.log(getX());
>   setX(6);
>   console.log(getX());
>   ```

- import가 실행되는 것 먼저인지, export의 평가 과정이 먼저인지 실행 순서는 정확하지는 않고 좀 더 공부가 필요하지만
- 모듈이 독자적인 스코프를 가지고 모듈에서 export 되는 함수가 클로저가 된다고 학습한 기억이 있음
- 따라서 export가 먼저 평가 및 실행이 될 것이라 추측하고 이어서 설명을 하면
- 메인 파일에서 getX가 호출되면 x는 상위 레시컬 환경 myModule 파일의 지역 변수 x에 할당된 x가 반환이 되어 5가 출력되고,
- setX함수를 호출하면 myModule 파일의 지역 변수 x에 6이 할당됨
- 다음 getX를 다시 호출하면 변경된 값 6이 참조되어 출력됨

**Comment by pipisebastian** [comment]
> * 아래 모듈 코드가 평가, 실행되는 과정을 설명해주세요.
>   ```js
>   // myModule.js
>   let x = 5;
>   export const getX = () => x;
>   export const setX = (val) => {
>     x = val;
>   };
>   
>   // --- 파일 분리 ---
>   
>   import { getX, setX } from "./myModule.js";
>   
>   console.log(getX());
>   setX(6);
>   console.log(getX());
>   ```

- `main` 평가+실행 →  `myModule` import 문 실행
- `myModule` 코드 평가(x 선언, getX, setX 함수 평가) + 실행(x 할당) → `main` 으로 돌아감
- `main` getX 호출 → x 는 상위 스코프에서 참조하여, 5 반환
- `main` setX 호출 → x 에 6 할당
- `main` getX 호출 → 6 반환
